engine/src/test/java/org/camunda/bpm/engine/test/errorcode/conf/CustomErrorCodeProviderTest.java

```diff
@@ -0,0 +1,352 @@
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+
+import java.sql.SQLException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.junit.Assert.fail;
+
+public class CustomErrorCodeProviderTest {
+
+  protected static int PROVIDED_CUSTOM_CODE = 33_333;
+
+  @ClassRule
+  public static ProcessEngineBootstrapRule bootstrapRule = new ProcessEngineBootstrapRule(c -> {
+    c.setCustomExceptionCodeProvider(new ExceptionCodeProvider() {
+
+      @Override
+      public Integer provideCode(SQLException sqlException) {
+        return PROVIDED_CUSTOM_CODE;
+      }
+
+      @Override
+      public Integer provideCode(ProcessEngineException processEngineException) {
+        return PROVIDED_CUSTOM_CODE;
+      }
+
+    });
+  });
+
+  @Rule
+  public ProcessEngineLoggingRule loggingRule = new ProcessEngineLoggingRule()
+      .watch("org.camunda.bpm.engine.cmd")
+      .level(Level.WARN);
+
+  protected ProvidedProcessEngineRule engineRule = new ProvidedProcessEngineRule(bootstrapRule);
+  protected ProcessEngineTestRule testRule = new ProcessEngineTestRule(engineRule);
+
+  @Rule
+  public RuleChain ruleChain = RuleChain.outerRule(engineRule).around(testRule);
+
+  protected RuntimeService runtimeService;
+  protected IdentityService identityService;
+
+  @Before
+  public void assignServices() {
+    runtimeService = engineRule.getRuntimeService();
+    identityService = engineRule.getIdentityService();
+  }
+
+  @After
+  public void clear() {
+    engineRule.getIdentityService().deleteUser("kermit");
+  }
+
+  @Test
+  public void shouldOverrideProvidedExceptionCode1() {
+    // given
+    BpmnModelInstance myProcess = Bpmn.createExecutableProcess("foo")
+        .startEvent()
+        .serviceTask()
+          .camundaClass(FailingJavaDelegateWithErrorCode.class)
+        .endEvent()
+        .done();
+
+    testRule.deploy(myProcess);
+
+    try {
+      // when
+      runtimeService.startProcessInstanceByKey("foo",
+          Variables.putValue("code", 22_222));
+      fail("Exception expected");
+
+    } catch (ProcessEngineException e) {
+      // then
+      assertThat(e.getCode()).isEqualTo(22_222);
+
+    }
+  }
+
+  @Test
+  public void shouldOverrideProvidedExceptionCode2() {
+    // given
+    BpmnModelInstance myProcess = Bpmn.createExecutableProcess("foo")
+        .startEvent()
+        .serviceTask()
+          .camundaClass(FailingJavaDelegateWithErrorCode.class)
+        .endEvent()
+        .done();
+
+    testRule.deploy(myProcess);
+
+    try {
+      // when
+      runtimeService.startProcessInstanceByKey("foo",
+          Variables.putValue("code", 20_000));
+      fail("Exception expected");
+
+    } catch (ProcessEngineException e) {
+      // then
+      assertThat(e.getCode()).isEqualTo(20_000);
+
+    }
```

Comment:
In some tests, I wanted to ensure that the getter is called. With `assertThatThrownBy` combined with `extracting`, I cannot check that since it retrieves the value via reflection from the field. I now changed it to the `catchThrowable` approach since this is more AssertJ-like.

Reply:
OK, I didn't know that, good to know 👍